Provide BOOST_ROOT to CMake docs target (RIPD-1364)

ximinez opened the pull request:
* Should make building docs with CMake incrementally easier and more
reliable.

I've named more reviewers than usual, because they are known to have different environments than I do, and I'd like to see if this is any easier. These changes only affect the [CMake](https://www.cmake.org) build.

If you don't already have a CMake build folder set up:
```
mkdir build/cmake # or whatever name you'd prefer
pushd build/cmake
cmake ../.. #  -G"Visual Studio 14 2015 Win64" if on Windows
popd
```
Once you have the folder set up, try to build the docs
```
cmake --build <cmake build folder> --target docs
```
Please report back your results and any problems you run into along the way.

ximinez commented on CMakeLists.txt:
```diff
@@ -440,15 +440,32 @@ set_property(TARGET ${other_target} PROPERTY EXCLUDE_FROM_DEFAULT_BUILD true)
   B2_EXE
   NAMES b2
-  PATHS ENV BOOST_ROOT
+  PATHS ${BOOST_ROOT}
```
Will add `HINTS`.

mellery451 commented:
👍  works well for me

bachase commented:
On my windows box, I have `BOOST_ROOT` set but am getting

```
$ cmake ../.. -G"Visual Studio 14 2015 Win64"
-- Could NOT find Boost
CMake Warning at Builds/CMake/CMakeFuncs.cmake:299 (message):
  Boost directory found, but not all components.  May not be able to build.
Call Stack (most recent call first):
  CMakeLists.txt:136 (use_boost)
```

When I try to run 
```
cmake --build <cmake build folder> --target docs
```

`makeqbk.sh` pops open in visual studio code and doesn't actually run.  That sounds like my problem though.

ximinez commented:
Rebased on 0.40.0.

bachase commented:
@ximinez My issues were all environment related.  Rebuilding boost and fixing the association for sh files fixes everything.

ximinez commented:
@bachase Because I don't like being dependent on Windows file association settings, I just pushed a change which will call `makeqbk` via `bash` if possible. Please try to break it. (As a bonus, `makeqbk` will no longer run in a separate shell window.)

bachase commented:
I reverted the Windows file association and tried your changes.  It seems to find bash ok, but bash can't find `makeqbk.sh`! I added a CMake `MESSAGE` statement to print `CMAKE_SOURCE_DIR` and it is properly set to the root of my clone.  

```
/usr/bin/bash: makeqbk.sh: No such file or directory
C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.CppCommon.targets(171,5): error MSB6006: "cmd.exe" exited with code 127. [C:\code\rippled_review\build\cmake\docs.vcxproj]
```

scottschurr commented:
Apparently I'm missing something?  I invoke the documentation build like this:
```
cd ./build/cmake/${BUILD_TYPE}
cmake --build . --target docs
cd ../../..
```
The documentation build plugs along okay for quite a while, but generates this error:
```
...
Generating XML output for namespace ripple
Generating XML output for file STObject.h
Generate XML output for dir /Users/ScottsMini/Development/rippled/src/ripple/protocol/
Generate XML output for dir /Users/ScottsMini/Development/rippled/src/ripple/
lookup cache used 293/65536 hits=7144 misses=297
finished...
warning: failed to load external entity "docca/include/docca/doxygen.xsl"
compilation error: file reference.xsl line 11 element include
xsl:include : unable to load docca/include/docca/doxygen.xsl
make[3]: *** [CMakeFiles/docs] Error 5
make[2]: *** [CMakeFiles/docs.dir/all] Error 2
make[1]: *** [CMakeFiles/docs.dir/rule] Error 2
make: *** [docs] Error 2
```
Do you know where I should be finding `doxygen.xsl`?

scottschurr commented:
My `ripple-libpp` build suggests I should be finding `doxygen.xls` in `docs/docca/include/docca/doxygen.xsl`.  I have a `docs/docca` directory, but it is empty.  Any ideas what I messed up?

mellery451 commented:
@scottschurr   `git submodule update --init`  ?

scottschurr commented:
Thanks @mellery451, that did the trick.  After that fix the documentation build worked great.

ximinez commented:
@scottschurr Just to save a bit of typing, once you've generated the cmake files the first time, you can build using just `cmake --build build/cmake/${BUILD_TYPE} --target docs`.

scottschurr commented:
@ximinez Thanks.  Yeah, even that is too much for me to remember, so I made a script named `cmk_docs_debug` that keeps the details hidden from me unless I need to know what those details are.  I've tidied up the script with your suggestion.

Would you like me to re-test after this most recent commit?

ximinez commented:
@scottschurr I run a lot of my common commands by back-searching my shell history, so I get it. And, yes, I'd appreciate if you could test the most recent change, too.

scottschurr commented:
My doc building script still works with the most recent "[FOLD]" commit.  Still :+1:

ximinez commented:
Rebased on 0.50.0-b1 (twice, because I screwed it up the first time) and squashed. I'm going to mark this as passed, since it has two 👍 s, but I'd still appreciate if @HowardHinnant and @JoelKatz could give it a shot to see if it works for them.

HowardHinnant commented:
@mellery451 I have the same symptoms as Scott. From what directory do I `git submodule update --init`?

ximinez commented:
It should work anywhere in the git source tree, but I usually just go from
the top.

On Tue, Dec 27, 2016, 3:12 PM Howard Hinnant <notifications@github.com>
wrote:

> @mellery451 <https://github.com/mellery451> I have the same symptoms as
> Scott. From what directory do I git submodule update --init?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ripple/rippled/pull/1937#issuecomment-269376700>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGcdUosA-tt-wcdUXJVcnHjFr5x3VZuVks5rMXE6gaJpZM4LRUpL>
> .
>
-- 

*Edward Hennis*Senior Software Engineer | Ripple
@EdwardHennis
PGP: F2198E3A / 5294 9650 7465 B591 A982  831C D835 08C0 F219 8E3A
ed@ripple.com | ripple.com

HowardHinnant commented:
👍  Works for me on macOS.

bachase commented:
Merged as a1c0d15a1ffd5d5ee665a2996115e4f5f73a5219
